packages/contracts-bedrock/src/L1/StandardValidator.sol

```diff
@@ -0,0 +1,271 @@
+import { IOptimismPortal2 } from "interfaces/L1/IOptimismPortal2.sol";
+import { IPreimageOracle } from "interfaces/cannon/IPreimageOracle.sol";
+
+contract StandardValidatorV180 {
+    ISuperchainConfig internal superchainConfig;
+    address internal l1PAOMultisig;
+    address internal mips;
+
+    // Semver strings for our contracts. Pulled from the Superchain Registry.
+
+    string internal constant systemConfigVersion = "2.3.0";
+    string internal constant permissionedDisputeGameVersion = "1.3.1";
+    string internal constant mipsVersion = "1.2.1";
+    string internal constant optimismPortalVersion = "3.10.0";
+    string internal constant anchorStateRegistryVersion = "2.0.0";
+    string internal constant delayedWETHVersion = "1.1.0";
+    string internal constant disputeGameFactoryVersion = "1.0.0";
+    string internal constant l1CrossDomainMessengerVersion = "2.3.0";
+    string internal constant l1ERC721BridgeVersion = "2.1.0";
+    string internal constant l1StandardBridgeVersion = "2.1.0";
+    string internal constant optimismMintableERC20FactoryVersion = "1.9.0";
+    string internal constant preimageOracleVersion = "1.1.2";
+
+    /// Bytecode hashes of our contracts. Does not include contracts that contain immutables. Pulled from
+    /// the Superchain Registry.
+
+    bytes32 internal constant systemConfigHash = 0x165dece6291f3a071e602c804f4a3e89977a71ec866f46c85eb21448f22f6f9a;
+    bytes32 internal constant permissionedDisputeGameHash =
+        0x1a81c5fe9fbbf68a0f75e987451e5181c4fc3e801f5c056d38f08942cefed349;
+    bytes32 internal constant mipsHash = 0xd80e53cfb3a0655e5a82b40f370b7352bd926d6d28de7f1735583fcfeba3f0fd;
+    bytes32 internal constant optimismPortalHash = 0x2b0610230634fccf9d05f4aaefceffef9e04aa99910798d42823e0068eb07eca;
+    bytes32 internal constant anchorStateRegistryHash =
+        0x165dece6291f3a071e602c804f4a3e89977a71ec866f46c85eb21448f22f6f9a;
+    bytes32 internal constant delayedWETHHash = 0x1a81c5fe9fbbf68a0f75e987451e5181c4fc3e801f5c056d38f08942cefed349;
+    bytes32 internal constant disputeGameFactoryHash =
+        0xd80e53cfb3a0655e5a82b40f370b7352bd926d6d28de7f1735583fcfeba3f0fd;
+    bytes32 internal constant l1CrossDomainMessengerHash =
+        0xc919b535f7bbbaa57f32b602f26000e3997c22c933efbb92a36e053b20b2b4b1;
+    bytes32 internal constant l1ERC721BridgeHash = 0xa1a3ea3996797cf5035dafff9335386f8b00c58ff22aaa4c34b02c4ae7cf5e8a;
+    bytes32 internal constant l1StandardBridgeHash = 0x6894839591ba2aa1f6ba43526113e25fd795ec5c5a48202e664c931cd170aaee;
+    bytes32 internal constant optimismMintableERC20FactoryHash =
+        0xb8ec72533005496795f8fc0c71f0312578ff626f7b94699984e1add6dc2f3b88;
+    bytes32 internal constant preimageOracleHash = 0x2109eb20fb34703fe7ab0231bf0e93d972fe78423b72e7fb1f797f168217b87d;
+
+    address internal constant defaultL2Sender = 0x000000000000000000000000000000000000dEaD;
+
+    bool internal isFailed = false;
+
+    string internal collectedErrors = "";
```

Comment:
I think it will be much cheaper to define this as `string memory collectedErrors = ""` directly in the `validate` method and pass that error string around, because writing to storage is very expensive and will not get fully refunded. To benchmark this you can use the `vm.lastCallGas()` cheat within a test to see what the cost is before and after this change